Library/Homebrew/cmd/uninstall.rb

```diff
@@ -26,6 +53,36 @@ def uninstall
         name = Formulary.canonical_name(name)
         rack = HOMEBREW_CELLAR/name
 
+        # Cellar/oldname can be a symlink to Cellar/newname so in order not to
+        # get newname's keg as oldname's keg we check if it's a symlink.
+        if rack.symlink?
+          raise "#{rack.to_s} is a symlink"
+        end
```

Comment:
My only fear is that we can refer to `newname` using `oldname` everywhere we instatiate a formula. There are a lot of such places in the code. With previous implementation we could get `newname` formula by its name *ONLY* in `Dependency#to_formula` and everywhere we called `from_oldname`.